cray_product_catalog/util/merge_dict.py

```diff
@@ -0,0 +1,156 @@
+# MIT License
+#
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# Contains a utility function for merging two dictionaries together.
+
+from copy import deepcopy
+
+
+def _dict_contains_no_subdicts_or_lists(dict_to_check):
+    """Return true if the given dict has no list or dictionary values.
+
+    Helper for merge_dict.
+
+    Args:
+        dict_to_check (dict): The dict to check for list/dictionary values.
+
+    Returns:
+        bool: True if the dict has no dict or list values.
+    """
+    return not any(isinstance(value, dict) or isinstance(value, list)
+                   for value in dict_to_check.values())
+
+
+def _values_are_dicts(*values):
+    """Return true if all given values are dict type.
+
+    Helper for merge_dict.
+
+    Args:
+        *values: A list of values to check.
+
+    Returns:
+        bool: True if the values are dicts.
+    """
+    return all(isinstance(v, dict) for v in values)
+
+
+def _values_are_lists(*values):
+    """Return true if all values are list type.
+
+    Helper for merge_dict.
+
+    Args:
+        *values: A list of values to check.
+
+    Returns:
+        bool: True if the values are lists.
+    """
+    return all(isinstance(v, list) for v in values)
+
+
+def _values_are_different_types(first_value, second_value):
+    """Return true if two values are not the same type.
+
+    Helper for merge_dict.
+
+    Args:
+        first_value: the value to compare with second_value.
+        second_value: the value to compare with first_value.
+
+    Returns:
+        bool: True if the two values are not the same type.
+    """
+    return not (
+        isinstance(first_value, type(second_value)) or
+        isinstance(second_value, type(first_value))
```

Comment:
The second isinstance is unnecessary.

Reply:
I wanted to handle (admittedly pretty edge-like) cases where we might be dealing with a subclass of a dictionary.
```
>>> class MySpecialDict(dict):
...   pass
...
>>> dictA = MySpecialDict()
>>> dictB = {}
>>> isinstance(dictA, type(dictB))
True
>>> isinstance(dictB, type(dictA))
False
```

In this case I think I would like the utility to consider a `MySpecialDict` and a regular `dict` to be the same type regardless of the order of the arguments.

Reply:
I'm going to resolve conversation this because at worst I think it is just a potentially unnecessary/overkill line of code.

Reply:
Maybe we shouldn't allow subclasses to compare as the same. I think that would mean making the `or` into an `and`.

Reply:
Thinking about it some more, I'm actually going to go with that and `and` the two comparisons together. I don't think we should be allowing subclasses and their parent class to be considered the same type for the purposes of this utility. Realistically I don't see how this would matter if we're just dealing with parsed yaml structures, but I can't guarantee that this utility won't be used for something different in the future.